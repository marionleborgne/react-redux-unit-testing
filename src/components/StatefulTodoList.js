import React from 'react';
import TodoList from './TodoList';

export const defaultItem = {
  text: 'Create pure React example',
  id: 'default',
  done: true
};

export default class StatefulTodoList extends React.Component {

  state = { items: [defaultItem] };

  _onItemClick = clickedItem => this.setState({
    items: this.state.items.map(
      item => item.id === clickedItem.id ? { ...item, done: !item.done } : item
    )
  });

  _createItem = text => {
    const newItem = {
      text,
      id: Date.now(),
      done: false
    };
    this.setState(state => ({
      items: state.items.concat(newItem)
    }));
  };

  render() {
    return <TodoList
      onItemClick={this._onItemClick}
      createItem={this._createItem}
      items={this.state.items}
    />
  }
}
