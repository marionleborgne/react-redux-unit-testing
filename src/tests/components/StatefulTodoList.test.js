import React from 'react';
import { shallow, mount } from 'enzyme';
import StatefulTodoList, { defaultItem } from '../../components/StatefulTodoList';

describe('StatefulTodoList', () => {

  it('renders', () => {
    const wrapper = shallow(<StatefulTodoList />);
    expect(wrapper).toMatchSnapshot();
  });

  it('has the expected initial state', () => {
    const wrapper = mount(<StatefulTodoList />);
    const actual = wrapper.state().items;
    const expected = [defaultItem];
    expect(actual).toEqual(expected);
  });

  it('renders one TodoList', () => {
    const wrapper = shallow(<StatefulTodoList />);
    expect(wrapper.find('TodoList')).toHaveLength(1);
  });

  it('creates a new item', () => {
    const wrapper = shallow(<StatefulTodoList />);
    expect(wrapper.state().items.length).toEqual(1);

    // Create a new item
    const newItem = { text: 'Test', done: false };
    wrapper.instance()._createItem(newItem.text);

    // Check that there is a new item in addition
    // to the default item.
    const actual = wrapper.state().items;
    expect(actual.length).toEqual(2);
    expect(actual[0]).toEqual(defaultItem);
    expect(actual[1].text).toEqual(newItem.text);
    expect(actual[1].done).toEqual(newItem.done);
    // We can't know what the default ID will be in advance,
    // but we know it should be defined.
    expect(actual[1].id).toBeDefined();
  });

  it('toggles an item', () => {
    const wrapper = shallow(<StatefulTodoList />);
    const initialItems = wrapper.state().items;
    expect(initialItems.length).toEqual(1);
    expect(initialItems[0]).toEqual(defaultItem);
    expect(initialItems[0].done).toEqual(defaultItem.done);

    // Toggle the status of the first item
    wrapper.instance()._onItemClick(initialItems[0]);

    // Check that there is a new item in addition
    // to the default item.
    const actual = wrapper.state().items;
    expect(actual.length).toEqual(1);
    expect(actual[0].done).toEqual(!defaultItem.done);
    expect(actual[0].text).toEqual(defaultItem.text);
    expect(actual[0].id).toEqual(defaultItem.id);
  });
});
