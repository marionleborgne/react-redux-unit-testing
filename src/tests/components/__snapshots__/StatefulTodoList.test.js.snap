// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatefulTodoList renders 1`] = `
<TodoList
  createItem={[Function]}
  items={
    Array [
      Object {
        "done": true,
        "id": "default",
        "text": "Create pure React example",
      },
    ]
  }
  onItemClick={[Function]}
/>
`;
